units/github_status_api.go

```diff
@@ -0,0 +1,316 @@
+	"github.com/evergreen-ci/evergreen/model/build"
+	"github.com/evergreen-ci/evergreen/model/patch"
+	"github.com/google/go-github/github"
+	"github.com/mongodb/amboy"
+	"github.com/mongodb/amboy/job"
+	"github.com/mongodb/amboy/registry"
+	"github.com/mongodb/grip"
+	"github.com/mongodb/grip/message"
+	"github.com/mongodb/grip/sometimes"
+	"github.com/pkg/errors"
+	"golang.org/x/oauth2"
+)
+
+const (
+	githubStatusUpdateJobName = "github-status-update"
+	githubStatusError         = "error"
+	githubStatusFailure       = "failure"
+	githubStatusPending       = "pending"
+	githubStatusSuccess       = "success"
+
+	githubUpdateTypeBuild = "build"
+	githubUpdateTypePatch = "patch"
+)
+
+func init() {
+	registry.AddJobType(githubStatusUpdateJobName, func() amboy.Job { return makeGithubStatusUpdateJob() })
+}
+
+type githubStatus struct {
+	owner    string
+	repo     string
+	prNumber int
+	ref      string
+
+	urlPath     string
+	description string
+	context     string
+	ghStatus    string
+}
+
+func (status *githubStatus) Valid() bool {
+	if status.owner == "" || status.repo == "" || status.prNumber == 0 ||
+		status.ref == "" || status.urlPath == "" || status.context == "" {
+		return false
+	}
+	if status.ghStatus != githubStatusError && status.ghStatus != githubStatusFailure &&
+		status.ghStatus != githubStatusPending && status.ghStatus != githubStatusSuccess {
+		return false
+	}
+
+	return true
+}
+
+type githubStatusUpdateJob struct {
+	job.Base `bson:"job_base" json:"job_base" yaml:"job_base"`
+	env      evergreen.Environment
+
+	FetchID    string `bson:"fetch_id" json:"fetch_id" yaml:"fetch_id"`
+	UpdateType string `bson:"update_type" json:"update_type" yaml:"update_type"`
+}
+
+func makeGithubStatusUpdateJob() *githubStatusUpdateJob {
+	return &githubStatusUpdateJob{
+		env: evergreen.GetEnvironment(),
+		Base: job.Base{
+			JobType: amboy.JobType{
+				Name:    githubStatusUpdateJobName,
+				Version: 0,
+				Format:  amboy.BSON,
+			},
+		},
+	}
+}
+
+// NewGithubStatusUpdateJobForBuild creates a job to update github's API from a Build.
+// Status will be reported as 'evergreen-[build variant name]'
+func NewGithubStatusUpdateJobForBuild(buildID string) amboy.Job {
+	job := makeGithubStatusUpdateJob()
+	job.FetchID = buildID
+	job.UpdateType = githubUpdateTypeBuild
+
+	job.SetID(fmt.Sprintf("%s:%s-%s-%d", githubStatusUpdateJobName, job.UpdateType, buildID, time.Now()))
+	return job
+}
+
+func repoReference(owner, repo string, prNumber int, ref string) string {
+	return fmt.Sprintf("%s/%s#%d@%s", owner, repo, prNumber, ref)
+}
+
+// NewGithubStatusUpdateForPatch creates a job to update github's API from a
+// Patch. Status will be reported as 'evergreen'
+func NewGithubStatusUpdateJobForPatch(version string) amboy.Job {
+	job := makeGithubStatusUpdateJob()
+	job.FetchID = version
+	job.UpdateType = githubUpdateTypePatch
+
+	job.SetID(fmt.Sprintf("%s:%s-%s-%d", githubStatusUpdateJobName, job.UpdateType, version, time.Now()))
```

Comment:
we might want to include the string form of the unix time, as the default format of time.Time includes spaces, which is weird, and has too much detail.

awkwardly, if we put this in a remote/shared queue, we could have collisions because these jobs could have the same name

Reply:
The collisions thing is really unfortunate. Do we have a GUID generation function somewhere?